### DISCUSSION TITLE
🔎 by emoji

### DISCUSSION THREAD

**Discussion by psivesely** [signalstickers/signalstickers#855] [Ideas / feature request]
In Signal you can search stickers by emoji. Recently I realized among my hundreds of sticker packs, none of them contain stickers for 🚌 🚍🚊🚞🚂🚉🚆🚄🚇🚟🚝🚋🚈🛤🚎. How am I supposed to converse with my fellow public transit stans without the appropriate signal stickers??!

Searching for the relevant keywords “bus”, “train,” “transit” are returning some results, but they’re mostly too specific (viz., to municipal transit systems). I think that adding search by emoji could help me find the train with a🦄 conductor sticking their head out the window sticker I know must be out there!

**Comment by Ploonet** [comment]
Would love this. If the maintainer is still active, I'd be down to help implement this.

**Reply by romainricard** [reply]
I'm still active :)

This feature could be intersting. If you like, you can submit a PR.
I advise that you tap into the search system, and use attributes, e.g. `emoji: 🫠` , to stay in line with the current operations.

**Reply by Ploonet** [reply]
Thanks !

I've looked around the repos and it seemed that I'll need to touch on the `backend` repo as well. I'm thinking we could add a field to the `api/packs` response that contains a list of all unique emojis used in the pack. Example : `["🚌", "🚍",  "🚊", "🚞"]`

Then we rely on the frontend to filter the cached results like you already seem to be doing.

What do you think ? Am I understanding the architecture correctly ?

**Reply by romainricard** [reply]
The data for packs displayed on signalstickers come from 2 sources: signalstickers' backend, and the pack object returned by Signal's API (cf. [types](https://github.com/signalstickers/signalstickers/blob/master/src/etc/types.ts))

Signalstickers' backend only returns data that's been added by signalstickers' contributors after the pack has been created on Signal side (Tags, NSFW bool, original bool, editor's choice, ...). The rest is embedded in the (encrypted) object returned by Signal (cf. [StickerPackManifest from @signalstickers/stickers-client](https://github.com/signalstickers/stickers-client/blob/master/src/etc/types.ts)).
So the emoji will be located in the object returned by Signal, not by the backend.

If you are curious, you can get more information about how stickers works in Signal [here](https://github.com/signalstickers/signalstickers-client/blob/master/STICKERS_INTERNALS.md).

**Reply by romainricard** [reply]
Another point: could you please be mindful of performances? We already use a less-than-ideal model and delegate the computing on the client side by making _the client_ requests Signal's API and index stickers.

Maybe don't index emojis when the packs are received, and only start indexing + caching when the user launches a query with `emoji:` ? This would avoid impacting the (vast) majority of users who use the app without filtering by emoji. What do you think?

**Reply by Ploonet** [reply]
Thanks a lot for the detailed suggestions. This is a bigger task than I originally thought but I'll give it a go.

**Reply by romainricard** [reply]
Thanks! Do not hesitate to ping me here if you have questions 🤝

**Reply by Ploonet** [reply]
Hi

Just so that i stop giving any false hopes : i underestimated the feature and overestimated my available time.. sorry about that.

I'd still like to tackle it someday, but not anytime soon. In the meantime, maybe our discussion will help someone else !
